crypto/src/lib.rs

```diff
@@ -547,37 +456,152 @@ impl PrivateKey {
+    #[derive(Debug, Display, Clone, Copy)]
+    #[display(fmt = "Algorithm not supported")]
+    pub struct NoSuchAlgorithm;
+
+    #[cfg(feature = "std")]
+    impl std::error::Error for NoSuchAlgorithm {}
+
+    /// Error which occurs when converting to/from `Multihash`
+    #[derive(Debug, Clone, Display)]
+    pub struct MultihashConvertError {
+        reason: String,
+    }
+
+    impl MultihashConvertError {
+        #[cfg(not(feature = "ffi_import"))]
+        pub(crate) const fn new(reason: String) -> Self {
+            Self { reason }
+        }
+    }
+
+    #[cfg(feature = "std")]
+    impl std::error::Error for MultihashConvertError {}
+
+    impl From<NoSuchAlgorithm> for MultihashConvertError {
+        fn from(_: NoSuchAlgorithm) -> Self {
+            Self {
+                reason: String::from("Digest function not supported"),
+            }
+        }
+    }
+
+    /// Error which occurs when parsing [`PublicKey`]
+    #[derive(Debug, Display, Clone)]
+    pub enum KeyParseError {
+        /// Decoding hex failed
+        #[display(fmt = "Failed to decode. {_0}: {_1}")]
+        Decode(String, hex::FromHexError),
+        /// Converting bytes to multihash failed
+        #[display(fmt = "Failed to convert. {_0}: {_1}")]
+        Multihash(String, MultihashConvertError),
+    }
+
+    #[cfg(feature = "std")]
+    impl std::error::Error for KeyParseError {}
+    /// Error when dealing with cryptographic functions
+    #[derive(Debug, Display)]
+    pub enum Error {
+        /// Returned when trying to create an algorithm which does not exist
+        #[display(fmt = "Algorithm doesn't exist")] // TODO: which algorithm
+        NoSuchAlgorithm,
+        /// Occurs during deserialization of a private or public key
+        #[display(fmt = "Key could not be parsed. {_0}")]
+        Parse(String),
+        /// Returned when an error occurs during the signing process
+        #[display(fmt = "Signing failed. {_0}")]
+        Signing(String),
+        /// Returned when an error occurs during key generation
+        #[display(fmt = "Key generation failed. {_0}")]
+        KeyGen(String),
+        /// Returned when an error occurs during digest generation
+        #[display(fmt = "Digest generation failed. {_0}")]
+        DigestGen(String),
+        /// Returned when an error occurs during creation of [`SignaturesOf`]
+        #[display(fmt = "`SignaturesOf` must contain at least one signature")]
+        EmptySignatureIter,
+        /// A General purpose error message that doesn't fit in any category
+        #[display(fmt = "General error. {_0}")] // This is going to cause a headache
+        Other(String),
+    }
+
+    #[cfg(feature = "std")]
+    #[cfg(not(feature = "ffi_import"))]
+    impl From<ursa::CryptoError> for Error {
+        fn from(source: ursa::CryptoError) -> Self {
+            match source {
+                ursa::CryptoError::NoSuchAlgorithm(_) => Self::NoSuchAlgorithm,
```

Comment:
I can make the type public if this helps.

Reply:
I think I have addressed this?